### PR TITLE
Archiving Cart Promotion

### PR DESCRIPTION
**Pull request by Wojdylak**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | #13475
| License         | MIT

Based on #15375, I have split this PR into "Archiving Cart Promotion" and "Archiving Catalog Promotion"
There  are two cases to consider:
- Priority: it cannot be null and is assigned a "gedmo:sortable-position". This means that when we archive cart promotion, it still  influences the position of all promotions.
- API changes: I have added added PATCH operations, which removes all null values from responses in promotions. So we need to decide whether to add `skip_null_values: false` across all operations or not?

### CODE REVIEW

File: features/cart/shopping_cart/cart_promotion_integration.feature
Comment (GSadee) [inline features/cart/shopping_cart/cart_promotion_integration.feature:1]: I would be for moving this scenario to `features/promotion/receiving_discount` or `features/promotion/applying_promotion_rules` and focusing only on not applying archived promotions as the scenarios for applying valid promotions already exist.
The name could be change to: `not_applying_archived_promotions` or sth similar

File: features/cart/shopping_cart/cart_promotion_integration.feature [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        And the store has a product "PHP T-Shirt" priced at "$100.00"
+        And there is a promotion "Christmas sale"
+        And this promotion gives "$10.00" discount to every order
+
+    @api @ui
+    Scenario: Promotion is applied to the cart when it is valid
+        When I add product "PHP T-Shirt" to the cart
+        Then I should be on my cart summary page
+        And I should be notified that the product has been successfully added
+        And my cart total should be "$90.00"
```
Comment (GSadee) [inline features/cart/shopping_cart/cart_promotion_integration.feature]: I would be for removing these scenarios and check here only archived promotions. You could also add a scenario for checking if already applied promotion will be withdrawn when the admin archives it

File: features/promotion/managing_promotions/archiving_promotions.feature [hunk 2]
```diff
@@ -0,0 +1,41 @@
+@managing_promotions
+Feature: Archiving promotions
+    In order to hide no longer available promotions from the list and customers' use
+    As a Administrator
```
Comment (GSadee) [inline features/promotion/managing_promotions/archiving_promotions.feature]:
```suggestion
    As an Administrator
```

File: features/promotion/managing_promotions/archiving_promotions.feature [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    @api @ui
+    Scenario: Archiving a promotion
+        When I browse promotions
+        And I archive the "Christmas sale" promotion
+        Then I should see a single promotion in the list
+        And I should see the promotion "New Year sale" in the list
+
+    @domain
+    Scenario: Archiving a promotion does not remove it form the database
```
Comment (GSadee) [inline features/promotion/managing_promotions/archiving_promotions.feature]:
```suggestion
    Scenario: Archiving a promotion does not remove it from the database
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/applicators.xml [hunk 4]
```diff
@@ -33,5 +33,10 @@
         <service id="Sylius\Bundle\ApiBundle\Applicator\ProductReviewStateMachineTransitionApplicatorInterface" class="Sylius\Bundle\ApiBundle\Applicator\ProductReviewStateMachineTransitionApplicator">
             <argument id="sm.factory" type="service" />
         </service>
+
+        <service id="sylius.api.applicator.archiving_promotion" class="Sylius\Bundle\ApiBundle\Applicator\ArchivingPromotionApplicator">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/applicators.xml:37]:
```suggestion
        <service id="sylius.api.archiving_promotion_applicator" class="Sylius\Bundle\ApiBundle\Applicator\ArchivingPromotionApplicator">
```
🤔

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 5]
```diff
@@ -410,5 +410,12 @@
             </argument>
             <tag name="api_platform.filter" />
         </service>
+
+        <service id="sylius.api.exists_filter.archived_at" parent="api_platform.doctrine.orm.exists_filter" public="true">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml:414]:
```suggestion
        <service id="sylius.api.archived_at_exists_filter" parent="api_platform.doctrine.orm.exists_filter" public="true">
```
🤔

File: tests/Api/Admin/PromotionsTest.php [hunk 6]
```diff
@@ -576,6 +576,56 @@ public function it_does_not_update_a_promotion_with_duplicate_locale_translation
...
+
+        $this->assertResponse(
+            $this->client->getResponse(),
+            'admin/promotion/archive_promotion',
+            Response::HTTP_OK,
+        );
+    }
+
+    /** @test */
+    public function it_restore_a_promotion(): void
```
Comment (GSadee) [inline tests/Api/Admin/PromotionsTest.php]:
```suggestion
    public function it_restores_a_promotion(): void
```

**Review by NoResponseMate — APPROVED** [review]

**Comment by NoResponseMate** [comment]
Thank you, @Wojdylak!

**Pull request merged**